File: src/main/java/org/samoxive/safetyjim/discord/processors/InviteLink.java
package org.samoxive.safetyjim.discord.processors;

import net.dv8tion.jda.core.Permission;
import net.dv8tion.jda.core.entities.Member;
import net.dv8tion.jda.core.entities.Message;
import net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent;
import net.dv8tion.jda.core.exceptions.InsufficientPermissionException;
import org.samoxive.safetyjim.database.DatabaseUtils;
import org.samoxive.safetyjim.discord.DiscordBot;
import org.samoxive.safetyjim.discord.DiscordShard;
import org.samoxive.safetyjim.discord.DiscordUtils;
import org.samoxive.safetyjim.discord.MessageProcessor;

public class InviteLink extends MessageProcessor {
    private String[] blacklistedHosts = { "discord.gg/" };
    // We don't want to censor users that can issue moderative commands
    private Permission[] whitelistedPermissions = {
            Permission.ADMINISTRATOR,
            Permission.BAN_MEMBERS,
            Permission.KICK_MEMBERS,
            Permission.MANAGE_ROLES,
            Permission.MESSAGE_MANAGE,
    };

    @Override
    public boolean onMessage(DiscordBot bot, DiscordShard shard, GuildMessageReceivedEvent event) {
        Message message = event.getMessage();
        Member member = event.getMember();
        for (Permission permission: whitelistedPermissions) {
            if (member.hasPermission(permission)) {
                return false;
            }
        }

        boolean processorEnabled = DatabaseUtils.getGuildSettings(bot, bot.getDatabase(), event.getGuild()).getInvitelinkremover();
        if (!processorEnabled) {
            return false;
        }

        String content = message.getContentRaw();

        boolean inviteLinkExists = false;
        for (String blacklistedHost: blacklistedHosts) {
            if (content.contains(blacklistedHost)) {
                inviteLinkExists = true;
            }
        }

        if (!inviteLinkExists) {
            return false;
        }

        try {
            message.delete().complete();
            DiscordUtils.sendMessage(event.getChannel(), "I'm sorry " + member.getAsMention() + ", you can't send invite links here.");
        } catch (InsufficientPermissionException e) {
            return false;
        } catch (Exception e) {
            return true;
        }

        return true;
    }
}

File: src/main/java/org/samoxive/safetyjim/discord/processors/InviteLink.kt
package org.samoxive.safetyjim.discord.processors

import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent
import org.samoxive.safetyjim.database.SettingsEntity
import org.samoxive.safetyjim.discord.*
import org.samoxive.safetyjim.tryhardAsync

private val blocklistedHosts = arrayOf("discord.gg/")
private const val ACTION_REASON = "Sending invite links"

fun isInviteLinkBlocklisted(str: String) = blocklistedHosts.map { str.contains(it) }.any { it }

class InviteLink : MessageProcessor() {
    override suspend fun onMessage(bot: DiscordBot, shard: DiscordShard, event: GuildMessageReceivedEvent, settings: SettingsEntity): Boolean {
        val message = event.message
        val member = event.member!!
        val channel = event.channel
        val guild = event.guild
        val selfUser = event.jda.selfUser
        val targetUser = event.author

        if (!settings.inviteLinkRemover) {
            return false
        }

        if (member.isStaff()) {
            return false
        }

        val content = message.contentRaw
        if (!isInviteLinkBlocklisted(content)) {
            return false
        }

        val expirationDate = settings.getInviteLinkRemoverActionExpirationDate()
        tryhardAsync {
            message.delete().await()
            executeModAction(settings.inviteLinkRemoverAction, guild, channel, settings, selfUser, targetUser, ACTION_REASON, expirationDate)
        }

        return true
    }
}
